### PR TITLE
Explore raw data and add multiple simulation runs

### PR DESCRIPTION
**Pull request by ajlee21**
This PR adds 2 new features:

1.  0_process_data.ipynb explores the raw data by plotting the raw gene expression data onto UMAP dimensions and coloring the samples (datapoints) by different metadata values to look at trends in the data.

This was a very helpful suggestion from the prelim committee to examine my data and especially to look for batch effects across experiments.

2.  1_detect_nonlinear_signal.ipynb allows us to run multiple simulations (using multiple random seeds) where we select random sets of As and Bs, generate a step function, determine the ability for VAE to reconstruct this signal.  This nb generates the data for multiple runs and next we need to visualize these multiple runs as a confidence interval.

3.  Update my .yml files to ensure reproducibility:  Pa_interpolate.yml is the exported conda environment.  Pa_interpolate_simple.yml is a manually edited yml file that is simplified for more robust use on different OS.

My next PR will also clean up my directories.

### CODE REVIEW

File: env/Pa_interpolate_simple.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  - conda-forge::r-dplyr=0.7.6
+  - conda-forge::r-tidyr=0.8.0
+  - conda-forge::r-readr=1.1.1
+  - conda-forge::r-readxl=1.1.0
+  - conda-forge::r-cowplot=0.9.3
+  - conda-forge::r-heatmap3=1.1.1
+  - conda-forge::r-ggrepel=0.8.0
+  - conda-forge::r-optparse=1.6.0
+  - pip:
+    - matplotlib==3.0.0
```
Comment (danich1) [inline env/Pa_interpolate_simple.yml:47]: Would move matplotlib and tqdm into the dependencies section with the conda-forge channel.

File: scripts/functions/encode_decode.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        4.  This computation is repeated for all samples 
+
+    output: 
+         1. encoded expression profile per sample
+         2. predicted expression profile per sample
+
+    """
+
+    # Output file
+    predict_file = os.path.join(out_dir,
```
Comment (danich1) [inline scripts/functions/encode_decode.py:78]: Just a reminder for the future: 
```python
predict_file = f"{out_dir}/vae_predicted_gene_exp_seed/{str(seed_input)}.txt"
```

File: scripts/functions/vae.py [hunk 3]
```diff
@@ -63,15 +24,61 @@
...
     """
 
+    # The below is necessary in Python 3.2.3 onwards to
+    # have reproducible behavior for certain hash-based operations.
+    # See these references for further details:
+    # https://docs.python.org/3.4/using/cmdline.html#envvar-PYTHONHASHSEED
+    # https://github.com/keras-team/keras/issues/2280#issuecomment-306959926
+    randomState = 123
+    import os
+    os.environ['PYTHONHASHSEED'] = '0'
```
Comment (danich1) [inline scripts/functions/vae.py:51]: why do you need to set the environment variable here? (Personal curiosity)

File: scripts/nbconverted/1_detect_nonlinear_signal.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+# 
+# **Question**: Can VAE detect nonlinear signal in simulated data?
+# 
+# **Hypothesis**: If we generate a gene expression dataset with a strong signal, step function, a nonlinear compression method (VAE) will be able to learn this signal and predict new gene expression patterns better compared to using a linear comparession method (PCA) and using no compression method (all the genes)
+#  
+# **Study design**:
+# 1. Input: Add step function to to Pa gene expression dataset:
+# 2. Approach: Train nonlinear (VAE) and linear (PCA) compression algorithms using this simulated data
+# 3. Evaluation: For each sample in the Pa dataset compare spearman corr(predicted expression, actual expression)
+# 4. Output: Boxplot of the correlation scores per sample
```
Comment (danich1) [inline scripts/nbconverted/1_detect_nonlinear_signal.py]: Is there a box plot for these multiple seed runs?

**Review by danich1 — APPROVED** [review]
Great PR. Most small code design comments.  Really curious as to why there is a spike in the val loss when running your simulations with seed 2. (1_detect_nonlinear_signal.ipynb) Any ideas?

File: env/Pa_interpolate_simple.yml (hunk 1 above)
Comment (ajlee21) [reply env/Pa_interpolate_simple.yml:47]: So I'm not as proficient in conda environments, but I originally did have matplotlib in the dependency section and I got a failure to setup the environment.  I moved it based on a blog post.  But very happy to sit down and learn more about how to better setup these environment files in the future

File: scripts/functions/encode_decode.py (hunk 2 above)
Comment (ajlee21) [reply scripts/functions/encode_decode.py:78]: Thanks.  This one is running py3.5 right now but I might switch it later

File: scripts/functions/vae.py (hunk 3 above)
Comment (ajlee21) [reply scripts/functions/vae.py:51]: I definitely don't know all the details for this.  This was also based on a blog post that I found as I was looking for ways to make multiple VAE training runs reproducible.  In general VAE training is known to yield slightly different models, but this one post said that this seemed to work: https://keras.io/getting-started/faq/#how-can-i-obtain-reproducible-results-using-keras-during-development

File: scripts/nbconverted/1_detect_nonlinear_signal.py (hunk 4 above)
Comment (ajlee21) [reply scripts/nbconverted/1_detect_nonlinear_signal.py]: There is eventually supposed to be boxplots created, but I wasn't able to get to it and didn't want the PR to get too big.

One boxplot will correspond to one random seed run, that correct.  I have updated my comment so its more clear.
